cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaAgentClient.java

```diff
@@ -0,0 +1,131 @@
+import io.strimzi.operator.common.Util;
+
+/**
+ * Creates HTTP client and interacts with Kafka Agent's REST endpoint
+ */
+class KafkaAgentClient {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaAgentClient.class.getName());
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private static final String BROKER_STATE_REST_PATH = "/v1/broker-state/";
+    private static final int BROKER_STATE_HTTPS_PORT = 8443;
+
+    private String namespace;
+    private Reconciliation reconciliation;
+    private String cluster;
+    private Secret clusterCaCertSecret;
+    private Secret coKeySecret;
+    private HttpClient httpClient;
+
+    KafkaAgentClient(Reconciliation reconciliation, String cluster, String namespace, Secret clusterCaCertSecret, Secret coKeySecret) {
+        this.reconciliation = reconciliation;
+        this.cluster = cluster;
+        this.namespace = namespace;
+        this.clusterCaCertSecret = clusterCaCertSecret;
+        this.coKeySecret = coKeySecret;
+    }
+
+    private HttpClient createHttpClient() throws GeneralSecurityException, IOException {
+
+        SSLContext sslContext = SSLContext.getInstance("TLSv1.3");
+        KeyStore trustStore = KeyStore.getInstance("PKCS12");
+        trustStore.load(new ByteArrayInputStream(
+                        Util.decodeFromSecret(clusterCaCertSecret, "ca.p12")),
+                new String(Util.decodeFromSecret(clusterCaCertSecret, "ca.password"), StandardCharsets.UTF_8).toCharArray());
+
+        String trustManagerFactoryAlgorithm = TrustManagerFactory.getDefaultAlgorithm();
+        TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(trustManagerFactoryAlgorithm);
+        trustManagerFactory.init(trustStore);
+        char[] keyPassword = new String(Util.decodeFromSecret(coKeySecret, "cluster-operator.password"), StandardCharsets.UTF_8).toCharArray();
+        KeyStore coKeyStore = KeyStore.getInstance("PKCS12");
+        coKeyStore.load(new ByteArrayInputStream(
+                        Util.decodeFromSecret(coKeySecret, "cluster-operator.p12")),
+                keyPassword
+        );
+
+        String keyManagerFactoryAlgorithm = KeyManagerFactory.getDefaultAlgorithm();
+        KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(keyManagerFactoryAlgorithm);
+        keyManagerFactory.init(coKeyStore, keyPassword);
+
+        sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), null);
+
+        return HttpClient.newBuilder()
+                .sslContext(sslContext)
+                .build();
+    }
+
+    protected String doGet(URI uri) {
+        try {
+            HttpRequest req = HttpRequest.newBuilder()
+                    .uri(uri)
+                    .GET()
+                    .build();
+
+            if (httpClient == null) {
+                httpClient = createHttpClient();
+            }
```

Comment:
What is the reason for not doing this in constructor? If there is some, might be good to add a comment with explanation.

Reply:
The reason I didn't it put this in the constructor was to handle the exceptions in getBrokerState(). However thinking about this again, we probably should put this in the constructor and throw the exceptions in KafkaRoller. We don't want to throw exceptions in KafkaRoller only when we actually attempt to connect to KafkaAgent for compatibility reason. However, we should be able to able initialise the client whether KafkaAgent is available or not.

Reply:
I am confused. Now the it's done in the constructor, do we still need this code here?

Reply:
ah good spot, I forgot to delete it here.